Nil pointer dereference errors when setting partial values for vault.RetryConfiguration

# Expected Behavior

We have the option of setting the custom RetryConfiguration in the client_options.go file via the `func WithRetryConfiguration(configuration RetryConfiguration) ClientOption()` function.

But, when using this function, what If we set only partial fields in the struct `type RetryConfiguration struct` that we would want to control like 
```go
client, err := vault.New(
vault.WithAddress(endpoint),
vault.WithRequestTimeout(30*time.Second),
vault.WithRetryConfiguration(vault.RetryConfiguration{
	RetryWaitMin: 2 * time.Second,
	RetryMax:     3,
	RetryWaitMax: 3 * time.Second,
}),
)
```

I would expect the HTTP calls to go through fine and the rest of fields of the `type RetryConfiguration struct` to be set to defaults if we aren't setting it similar to what `func DefaultConfiguration() ClientConfiguration` is doing.

Can the method `func WithRetryConfiguration()` not check the unset fields and set them to defaults? 

# Current Behavior

What is the current behavior?

Current snippet :

```go
client, err := vault.New(
vault.WithAddress(endpoint),
vault.WithRequestTimeout(30*time.Second),
vault.WithRetryConfiguration(vault.RetryConfiguration{
	RetryWaitMin: 2 * time.Second,
	RetryMax:     3,
	RetryWaitMax: 3 * time.Second,
}),
)
```

is throwing errors complaining about Nil pointer dereference errors as the remaining fields `CheckRetry, Backoff, ErrorHandler & Logger` are unset during the http call.


# Failure Information

Please include the version of Vault binary and the version of `vault-client-go` you're using.

vault-client-go-v0.3.3

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Setup the vault client with below snippet
```go
client, err := vault.New(
vault.WithAddress(endpoint),
vault.WithRequestTimeout(30*time.Second),
vault.WithRetryConfiguration(vault.RetryConfiguration{
	RetryWaitMin: 2 * time.Second,
	RetryMax:     3,
	RetryWaitMax: 3 * time.Second,
}),
)
```
2. Perform a vault call like:
```go
resp, err := client.System.ReadHealthStatus(ctx)
```
3. You should see errors like when the HTTP call is being made -
```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12dfdd8]
```

# Additional Information

Full Stack trace
```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12dfdd8]

goroutine 1 [running]:
github.com/hashicorp/go-retryablehttp.(*Client).Do(0xc0001218f0, 0xc0000b2288)
	/Users/rohit.appu/go/pkg/mod/github.com/hashicorp/go-retryablehttp@v0.7.4/client.go:641 +0x678
github.com/hashicorp/vault-client-go.(*Client).do(0xc000122780, 0x0?, 0x0?)
	/Users/rohit.appu/go/pkg/mod/github.com/hashicorp/vault-client-go@v0.3.3/client_requests.go:445 +0x8e
github.com/hashicorp/vault-client-go.(*Client).send(0xc000122780, {0x170dba8?, 0xc00009ff50?}, 0x0?, 0x0?, {0x0, 0x0, 0x0?}, {0x0, 0x0, ...})
	/Users/rohit.appu/go/pkg/mod/github.com/hashicorp/vault-client-go@v0.3.3/client_requests.go:403 +0x11c
github.com/hashicorp/vault-client-go.sendRequestParseResponse[...]({0x170db70, 0xc0000a2000?}, 0xc000122780?, {0x1637f7f, 0x3}, {0x163bde1, _}, {0x0, 0x0}, _, ...)
	/Users/rohit.appu/go/pkg/mod/github.com/hashicorp/vault-client-go@v0.3.3/client_requests.go:260 +0x336
github.com/hashicorp/vault-client-go.(*System).ReadHealthStatus(0xc0001229d0, {0x170db70, 0xc0000a2000}, {0x0?, 0x14e3b40?, 0x77359400?})
	/Users/rohit.appu/go/pkg/mod/github.com/hashicorp/vault-client-go@v0.3.3/api_system.go:2723 +0x1f1
jcredctl/internal/vault.verifyConnectivity({0x170db70?, 0xc0000a2000?}, {0x1707d00?, 0xc0001229d0?})
	/Users/rohit.appu/Documents/Repositories/is-cloudbees-core/jcredctl/internal/vault/vault_base.go:106 +0x4b
jcredctl/internal/vault.New({0x0?, 0x0?}, {0x7ff7bfeff3c4, 0x1a})
	/Users/rohit.appu/Documents/Repositories/is-cloudbees-core/jcredctl/internal/vault/vault_base.go:37 +0xeb
jcredctl/cmd.InfoCmd.func1(0xc00012a600?, {0x16383cd?, 0x2?, 0x2?})
	/Users/rohit.appu/Documents/Repositories/is-cloudbees-core/jcredctl/cmd/info.go:21 +0x47
github.com/spf13/cobra.(*Command).execute(0xc00012a600, {0xc0000c2720, 0x2, 0x2})
	/Users/rohit.appu/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0xc00012a000)
	/Users/rohit.appu/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/rohit.appu/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
main.main()
	/Users/rohit.appu/Documents/Repositories/is-cloudbees-core/jcredctl/main.go:12 +0x1f
```

--- a/client_configuration.go
+++ b/client_configuration.go
@@ -166,7 +166,7 @@ type RetryConfiguration struct {
 	// Default: 1000 milliseconds
 	RetryWaitMin time.Duration `env:"VAULT_RETRY_WAIT_MIN"`
 
-	// MaxRetryWait controls the maximum time to wait before retrying when
+	// RetryWaitMax controls the maximum time to wait before retrying when
 	// a 5xx or 412 error occurs.
 	// Default: 1500 milliseconds
 	RetryWaitMax time.Duration `env:"VAULT_RETRY_WAIT_MAX"`

--- a/client_option.go
+++ b/client_option.go
@@ -54,18 +54,34 @@ func WithRequestTimeout(timeout time.Duration) ClientOption {
 }
 
 // WithTLS configures the TLS settings in the base http.Client.
-func WithTLS(configuration TLSConfiguration) ClientOption {
+func WithTLS(tls TLSConfiguration) ClientOption {
 	return func(c *ClientConfiguration) error {
-		c.TLS = configuration
+		c.TLS = tls
 		return nil
 	}
 }
 
 // WithRetryConfiguration configures the internal go-retryablehttp client.
 // The library sets reasonable defaults for this setting.
-func WithRetryConfiguration(configuration RetryConfiguration) ClientOption {
+func WithRetryConfiguration(retry RetryConfiguration) ClientOption {
 	return func(c *ClientConfiguration) error {
-		c.RetryConfiguration = configuration
+		// if any of the required RetryConfiguration values are missing, use defaults
+		defaultRetryConfiguration := DefaultConfiguration().RetryConfiguration
+
+		if retry.CheckRetry == nil {
+			retry.CheckRetry = defaultRetryConfiguration.CheckRetry
+		}
+
+		if retry.Backoff == nil {
+			retry.Backoff = defaultRetryConfiguration.Backoff
+		}
+
+		if retry.ErrorHandler == nil {
+			retry.ErrorHandler = defaultRetryConfiguration.ErrorHandler
+		}
+
+		c.RetryConfiguration = retry
+
 		return nil
 	}
 }

--- a/client_test.go
+++ b/client_test.go
@@ -30,7 +30,7 @@ func Test_Client_Clone(t *testing.T) {
 
 	clone := client.Clone()
 
-	testEqual(t, client, clone)
+	assertEqual(t, client, clone)
 
 	assert.Equal(t, "http://test", clone.Configuration().Address)
 	assert.Equal(t, 30*time.Second, clone.Configuration().RequestTimeout)
@@ -40,8 +40,8 @@ func Test_Client_Clone(t *testing.T) {
 	assert.Equal(t, "test-namespace", clone.clientRequestModifiers.headers.namespace)
 }
 
-// We cannot compare the two objects directly since they have nested func pointers
-func testEqual(t *testing.T, c1 *Client, c2 *Client) {
+// assertEqual compares the two clients, accounting for nested func pointers
+func assertEqual(t *testing.T, c1 *Client, c2 *Client) {
 	assert.Equal(t, fmt.Sprintf("%v", c1.configuration), fmt.Sprintf("%v", c2.configuration))
 	assert.Equal(t, fmt.Sprintf("%v", c1.parsedBaseAddress), fmt.Sprintf("%v", c2.parsedBaseAddress))
 	assert.Equal(t, fmt.Sprintf("%v", c1.client), fmt.Sprintf("%v", c2.client))

--- a/client_test.go
+++ b/client_test.go
@@ -40,7 +40,8 @@ func Test_Client_Clone(t *testing.T) {
 	assert.Equal(t, "test-namespace", clone.clientRequestModifiers.headers.namespace)
 }
 
-// assertEqual compares the two clients, accounting for nested func pointers
+// assertEqual compares the two clients, accounting for the nested func pointers,
+// which are not comparable in Go (https://go.dev/ref/spec#Comparison_operators)
 func assertEqual(t *testing.T, c1 *Client, c2 *Client) {
 	assert.Equal(t, fmt.Sprintf("%v", c1.configuration), fmt.Sprintf("%v", c2.configuration))
 	assert.Equal(t, fmt.Sprintf("%v", c1.parsedBaseAddress), fmt.Sprintf("%v", c2.parsedBaseAddress))

--- a/client_test.go
+++ b/client_test.go
@@ -41,7 +41,7 @@ func Test_Client_Clone(t *testing.T) {
 }
 
 // assertEqual compares the two clients, accounting for the nested func pointers,
-// which are not comparable in Go (https://go.dev/ref/spec#Comparison_operators)
+// which are not comparable in Go (https://go.dev/ref/spec#Comparison_operators).
 func assertEqual(t *testing.T, c1 *Client, c2 *Client) {
 	assert.Equal(t, fmt.Sprintf("%v", c1.configuration), fmt.Sprintf("%v", c2.configuration))
 	assert.Equal(t, fmt.Sprintf("%v", c1.parsedBaseAddress), fmt.Sprintf("%v", c2.parsedBaseAddress))

--- a/client_test.go
+++ b/client_test.go
@@ -4,6 +4,7 @@
 package vault
 
 import (
+	"fmt"
 	"testing"
 	"time"
 
@@ -29,7 +30,7 @@ func Test_Client_Clone(t *testing.T) {
 
 	clone := client.Clone()
 
-	assert.Equal(t, client, clone)
+	testEqual(t, client, clone)
 
 	assert.Equal(t, "http://test", clone.Configuration().Address)
 	assert.Equal(t, 30*time.Second, clone.Configuration().RequestTimeout)
@@ -37,4 +38,14 @@ func Test_Client_Clone(t *testing.T) {
 	assert.Equal(t, 900*time.Millisecond, clone.Configuration().RetryConfiguration.RetryWaitMax)
 	assert.Equal(t, "test-token", clone.clientRequestModifiers.headers.token)
 	assert.Equal(t, "test-namespace", clone.clientRequestModifiers.headers.namespace)
+}
+
+// We cannot compare the two objects directly since they have nested func pointers
+func testEqual(t *testing.T, c1 *Client, c2 *Client) {
+	assert.Equal(t, fmt.Sprintf("%v", c1.configuration), fmt.Sprintf("%v", c2.configuration))
+	assert.Equal(t, fmt.Sprintf("%v", c1.parsedBaseAddress), fmt.Sprintf("%v", c2.parsedBaseAddress))
+	assert.Equal(t, fmt.Sprintf("%v", c1.client), fmt.Sprintf("%v", c2.client))
+	assert.Equal(t, fmt.Sprintf("%v", c1.clientWithRetries), fmt.Sprintf("%v", c2.clientWithRetries))
+	assert.Equal(t, fmt.Sprintf("%v", c1.clientRequestModifiers), fmt.Sprintf("%v", c2.clientRequestModifiers))
+	assert.Equal(t, fmt.Sprintf("%v", c1.replicationStates.states), fmt.Sprintf("%v", c2.replicationStates.states))
 }

